pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -144,23 +147,70 @@ type stringSliceFunc func(storage.Querier) ([]string, storage.Warnings, error)
 // the output of the stringSliceFunc to be sorted, as results of LabelValues
 // are not sorted.
 //
-// TODO: Consider running stringSliceFunc calls concurrently
+
+type stringSliceFuncResult struct {
+	Result   []string
+	Warnings storage.Warnings
+}
+
 func (m *mergeQuerier) mergeDistinctStringSlice(f stringSliceFunc) ([]string, storage.Warnings, error) {
+
+	g, _ := errgroup.WithContext(m.ctx)
```

Comment:
context returned by errgroup.WithContext should not be ignored, but used by goroutines to stop early when any goroutine returns error (eg. by passing to `stringSliceFunc`). Unfortunately Prometheus Querier interface currently doesn't make this easy :(

Reply:
> Unfortunately Prometheus Querier interface currently doesn't make this easy :(

The lack of context parameters in the storage interface is frustrating. I understand Prometheus didn't have a strong need for them but I think we should make a push upstream to add context parameters to the storage interface functions.

Reply:
While I have added a `context.Context` to the stringSliceFunc definition in a5b94abc6e002f6038d755fc1e9007316b3c3712. I am not making use of it. 

The only way I could really think of using it, would be by ignoring the result of the call all together, but I think that could lead some serious risk of DoS surface:

```
	done := make(chan struct{})
	go func() {
		output, warnings, err = f(ctx, q)
		close(done)
	}()
	select {
	case <-done:
		return output, warnings, err
	case <-ctx.Done():
		return nil, nil, ctx.Err()
	}
```

Are there any better approaches?

Reply:
In ff839dc859fb5ae2811d10fd609a5a564fc00fe1 I now make at least sure before I start a job with `concurrent.ForEach`, that the context is not cancelled.